### PR TITLE
Add support for images in Markdown previews

### LINKED ISSUE #13246: Support images in Markdown preview
### Check for existing issues

- [X] Completed

### Describe the feature

cannot display images in the tabs whenI open a markdown file with `markdown:open preview`
image

https://github.com/zed-industries/zed/assets/45729654/96635e7c-d4a2-4db0-8fbe-291a588d650b

macos:13.4.1
zed:0.139.3
### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by dovakin0007**
Closes #13246 

Release Notes:
added image preview for markdown 

image

Release Notes:

- Added support for images to the markdown preview

### CODE REVIEW

**Comment by dovakin0007** [comment]
> @dovakin0007 : Does it also work using the `img` tag?
> 
> `Image description`

no sir

**Comment by dovakin0007** [comment]
> @dovakin0007 : What is displayed in case of the `alt` message?

the alt text like zed gpui image doesn't have a way to display the alt text so for now I am displaying alt text  
image this text will be shown if image fails or unable to find the image in path 
image
else we display image

File: crates/markdown_preview/src/markdown_parser.rs [hunk 1]
```diff
@@ -855,6 +900,7 @@ mod tests {
         } else {
             panic!("Expected a paragraph");
         };
+        println!("{:?}", paragraph);
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_parser.rs:903]: This looks some debug printing that should be removed.

File: crates/markdown_preview/src/markdown_parser.rs [hunk 2]
```diff
@@ -300,10 +303,39 @@ impl<'a> MarkdownParser<'a> {
...
+                            ));
                         }
                     }
+                    if let Some(image) = image.clone() {
+                        let is_valid_image = match image.clone() {
+                            Image::Path {
+                                display_path,
+                                path: _,
+                            } => display_path.is_file(),
+                            Image::Web { url } => match isahc::get(url) {
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_parser.rs:318]: I don't think we want to use this isahc::get API here. Rather, GPUI already has an `ImageSource` type that you can pass directly to `img()`, that will make the HTTP request or read the file for you. Maybe just parse the image tag into this type and use it?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

File: crates/markdown_preview/src/markdown_parser.rs (hunk 2 above)
Comment (dovakin0007) [reply crates/markdown_preview/src/markdown_parser.rs:318]: yeah I will do that then and commit the changes

Comment (dovakin0007) [reply crates/markdown_preview/src/markdown_parser.rs:318]: > I don't think we want to use this isahc::get API here. Rather, GPUI already has an `ImageSource` type that you can pass directly to `img()`, that will make the HTTP request or read the file for you. Maybe just parse the image tag into this type and use it?
 yeah the reason why I am using isahc get is to see whether the image is valid as per markdown spec if the image fails to get loaded we have to render the all source as of now GPUI Image doesn't return an error or any kind of alt text if the image fails to get loaded that's why I am adding a check here for now. any suggestions are welcome 😄

**Comment by dovakin0007** [comment]
I want to raise one more issue it seems like the image is not rendering in same line 
image like the above example any ideas and contributions will be helpful here

**Comment by mikayla-maki** [comment]
Seems like there's some test failures as well. I'm going to mark this as draft for now :)

As for your issues with the layout, perhaps you need to add some more styling rules on the `img` perhaps using`block()` (equivalent to `display: block`)?

**Comment by bennetbo** [comment]
I haven't looked at the PR in the detail, but I assume that we will run into the same issues I encountered in #10565.
In particular there are these two issues:

1. Gpui cannot handle images in the middle of a paragraph, therefore the parser needs to split up items that contain images and render it as their own block (I implemented this in #10565), or we need to add support in gpui for this (seems like a lot of work though)
2. The `img` element needs better error handling, right now if a link is malformed or does not return an image, we fetch it over and over again. This wastes unnecessary resources and network bandwidth.

Rendering images in separate lines feels like a good enough workaround for now, but 2. definitely feels like a blocker to me.

**Comment by dovakin0007** [comment]
> Seems like there's some test failures as well. I'm going to mark this as draft for now, once things are ready to go feel free to mark it as ready :)
> 
> As for your issues with the layout, perhaps you need to add some more styling rules on the `img` perhaps using`block()` (equivalent to `display: block`)?

yeah that should work. regarding the layout we have to render the image within text StyledText can't handle images for now

**Comment by dovakin0007** [comment]
> I haven't looked at the PR in the detail, but I assume that we will run into the same issues I encountered in #10565. In particular there are these two issues:
> 
> 1. Gpui cannot handle images in the middle of a paragraph, I think there are two ways we can address this:
> 
> * Split up paragraph's that contain images and render each text and image as their own block (This is what I did in [markdown preview: Image support #10565](https://github.com/zed-industries/zed/pull/10565))
> * Add support for this in gpui (seems like a lot of work though)
> 
> 2. The `img` element needs better error handling, right now if a link is malformed or does not return an image, we fetch it over and over again. This wastes unnecessary resources and network bandwidth.
> 
> Rendering images in separate lines feels like a good enough workaround for now, but 2. definitely feels like a blocker to me.

yeah I am running into the same issues that you have mentioned.

**Comment by dovakin0007** [comment]
> I haven't looked at the PR in the detail, but I assume that we will run into the same issues I encountered in #10565. In particular there are these two issues:
> 
> 1. Gpui cannot handle images in the middle of a paragraph, I think there are two ways we can address this:
> 
> * Split up paragraph's that contain images and render each text and image as their own block (This is what I did in [markdown preview: Image support #10565](https://github.com/zed-industries/zed/pull/10565))
> * Add support for this in gpui (seems like a lot of work though)
> 
> 2. The `img` element needs better error handling, right now if a link is malformed or does not return an image, we fetch it over and over again. This wastes unnecessary resources and network bandwidth.
> 
> Rendering images in separate lines feels like a good enough workaround for now, but 2. definitely feels like a blocker to me.
Maybe we can split the text into multiple blocks which will allow us to render image between text that can fix this 2?

**Comment by dovakin0007** [comment]
still img() doesn't return error for bad request

**Comment by dovakin0007** [comment]
I was able to render the image within

Uploading markdown_preview_2.mp4…

 text

File: crates/markdown_preview/src/markdown_elements.rs [hunk 3]
```diff
@@ -267,3 +282,93 @@ impl Display for Link {
...
+                source_range,
+                display_path: path.clone(),
+                path,
+                link,
+            });
+        }
+        if let Some(file_location_directory) = file_location_directory {
+            let display_path = path;
+            let path = file_location_directory.join(text);
+            if path.exists() {
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_elements.rs:335]: Blocking IO

File: crates/markdown_preview/src/markdown_elements.rs [hunk 4]
```diff
@@ -267,3 +282,93 @@ impl Display for Link {
...
+    ) -> Option<Image> {
+        if text.starts_with("http") {
+            return Some(Image::Web {
+                source_range,
+                url: text,
+                link,
+            });
+        }
+        let path = PathBuf::from(&text);
+        if path.is_absolute() && path.exists() {
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_elements.rs:324]: We don't want to do blocking IO on the main thread.

File: crates/markdown_preview/src/markdown_parser.rs [hunk 5]
```diff
@@ -224,8 +223,54 @@ impl<'a> MarkdownParser<'a> {
...
+                                source_range: _,
+                                display_path,
+                                path: _,
+                                link: _,
+                            } => gpui::ImageSource::try_from(display_path).is_ok(),
+                            Image::Web {
+                                source_range: _,
+                                url,
+                                link: _,
+                            } => match isahc::get(url) {
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_parser.rs:239]: I think we should just skip this check. It's ok if the image is invalid on the web, just let GPUI fail to render it.

File: crates/markdown_preview/src/markdown_elements.rs [hunk 6]
```diff
@@ -107,7 +122,7 @@ pub enum ParsedMarkdownTableAlignment {
 #[derive(Debug)]
 #[cfg_attr(test, derive(PartialEq))]
 pub struct ParsedMarkdownTableRow {
-    pub children: Vec<ParsedMarkdownText>,
+    pub children: Vec<Vec<MarkdownParagraph>>,
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_elements.rs:125]: Why have these code paths been changed to `Vec<Vec<_>>`?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

File: crates/markdown_preview/src/markdown_elements.rs (hunk 6 above)
Comment (dovakin0007) [reply crates/markdown_preview/src/markdown_elements.rs:125]: I have changed it just to make it render text and image together it's a bit hacky solution but on most cases length of inner vector will be 1

**Comment by MatinAniss** [comment]
Support for inline elements in the text system is probably the better choice if the text and inline elements needs to wrap properly. Splitting the text and images into their own blocks is just avoiding the problem that will eventually will be needed to be implemented.

I've tried to implement it myself, but there are some changes and some decisions which I'm not sure what is best.

I have had a issue up to track this https://github.com/zed-industries/zed/issues/10916

**Comment by dovakin0007** [comment]
> Support for inline elements in the text system is probably the better choice if the text and inline elements needs to wrap properly. Splitting the text and images into their own blocks is just avoiding the problem that will eventually be needed to be implemented. It may be fine for now.
> 
> I've tried to implement it myself, but there are some changes and some decisions which I'm not sure what is best.
> 
> I have had a issue up to track this https://github.com/zed-industries/zed/issues/10916

Yeah I did try to do it but I wasn't on the right track so I scrapped it then for now implemented to render image as its separate own block rather than a part of interactive text yeah you are right once interactive text supports inline images this has to be changed

**Comment by dovakin0007** [comment]
@mikayla-maki I have made the changes please do review when you have time thanks

**Comment by dovakin0007** [comment]
> | A | B |
> |---|---|
> | A | B |

image

**Comment by dovakin0007** [comment]
> > A
> > B
> > 
> > 
> > 
> > 
> > A
> > B
> 
> image

the image will be misaligned due to https://github.com/zed-industries/zed/issues/11024

**Comment by dovakin0007** [comment]
> @dovakin0007 : This problem also had @bennetbo in #10565.

Yeah I am running into same issues as he mentioned

File: crates/markdown_preview/src/markdown_renderer.rs [hunk 7]
```diff
@@ -152,7 +153,7 @@ fn render_markdown_heading(parsed: &ParsedMarkdownHeading, cx: &mut RenderContex
         .text_color(color)
         .pt(rems(0.15))
         .pb_1()
-        .child(render_markdown_text(&parsed.contents, cx))
+        .child(div().children(render_markdown_text(&parsed.contents, cx)))
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_renderer.rs:156]: Can we just change this to a `children()` call?

```suggestion
        .children(render_markdown_text(&parsed.contents, cx))
```

File: crates/markdown_preview/src/markdown_renderer.rs [hunk 8]
```diff
@@ -270,7 +271,7 @@ fn render_markdown_table_row(
 
         let mut cell = container
             .w_full()
-            .child(contents)
+            .child(div().children(contents))
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_renderer.rs:274]: Same here

```suggestion
            .children(contents)
```

File: crates/markdown_preview/src/markdown_renderer.rs [hunk 9]
```diff
@@ -339,83 +340,152 @@ fn render_markdown_code_block(
...
+                let mut links = Vec::new();
+                let mut link_ranges = Vec::new();
+                for (range, region) in parsed.region_ranges.iter().zip(&parsed.regions) {
+                    if let Some(link) = region.link.clone() {
+                        links.push(link);
+                        link_ranges.push(range.clone());
+                    }
                 }
+                let workspace = cx.workspace.clone();
+                let x = div()
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_renderer.rs:394]: We should choose a better variable name than x

**Review by mikayla-maki — CHANGES_REQUESTED** [review]
A few nitpicks, then I'd be happy to merge this!

**Comment by dovakin0007** [comment]
> A few nitpicks, then I'd be happy to merge this!

Will fix those thanks !

**Comment by dovakin0007** [comment]
> @dovakin0007 : The problem with the tables has probably been solved in #18315. By updating with the latest commit of zed you can see if the problem with the images in the tables still persists or has been solved.

okay so it seems like there is a merge conflict due to     
`for (index, cell) in parsed.header.children.iter().enumerate() {
        let length = cell.contents.len();
        max_lengths[index] = length;
    }

    for row in &parsed.body {
        for (index, cell) in row.children.iter().enumerate() {
            let length = cell.contents.len();
            if length > max_lengths[index] {
                max_lengths[index] = length;
            }
        }
    }`

**Comment by dovakin0007** [comment]
> @dovakin0007 : The problem with the tables has probably been solved in #18315. By updating with the latest commit of zed you can see if the problem with the images in the tables still persists or has been solved.

image
it has been fixed

**Comment by dovakin0007** [comment]
> @dovakin0007 : What are the remaining problems?

Adjust table length based on image size issue is no way to get image length for now

**Comment by dovakin0007** [comment]
> @dovakin0007 : Are you also referring to the problem that the html logo image is not aligned correctly to the center like vscode and github does?
> 
> 
> 
> Vscode:
> 
> Screenshot 2024-10-01 alle 19 08 56
> 
> 

Yes and if I add it and the table width does change based on the image width

**Comment by dovakin0007** [comment]
> @dovakin0007 : Picking from something similar using the image dependency.
> 
> https://github.com/zed-industries/zed/blob/9b148f3dcc5e4281bfadd515efd817bcdbf21bc3/crates/gpui/src/assets.rs#L74-L78
yeah to  get the frame index some changes need to be done in GPUI for now I am gonna leave it as it is 
https://github.com/zed-industries/zed/blob/9cd42427d88afad3423fa2546ed656839295cf3f/crates/gpui/src/elements/img.rs#L164

**Comment by dovakin0007** [comment]
@mikayla-maki  I have done the changes let me know if anything else need to be changed

**Comment by dovakin0007** [comment]
Hi @mikayla-maki the conflict has been resolved let me know if anything else needed to be changed

**Comment by mjdNjhNJ** [comment]
@dovakin0007 Thank you for the work! I hope this gets merged soon.

**Comment by dovakin0007** [comment]
will re open once I add fallback text for images

**Comment by dovakin0007** [comment]
Hi everyone please do review the code and let me know what are the changes need to be done and feel free to change it

File: crates/markdown_preview/src/markdown_elements.rs [hunk 10]
```diff
@@ -35,6 +41,13 @@ impl ParsedMarkdownElement {
     }
 }
 
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum MarkdownParagraph {
```
Comment (mikayla-maki) [inline crates/markdown_preview/src/markdown_elements.rs:46]: Can we change this to "MarkdownParagraphChunk"?

And the names of these types to just "Text" and "Image"? Perhaps with a `type MarkdownParagraph = Vec<MarkdownParagraphChunk>`? This way it's clear why markdown paragraphs are represented as a Vec<Vec<MarkdownParagraphChunk>>

**Review by mikayla-maki — APPROVED** [review]
Thanks for your patience on this! I think this PR looks ready to go, I pushed some code style nits to a different PR here: https://github.com/zed-industries/zed/pull/21082, so I'm going to close this one for now. But it's all the same code and commits, thank you for implementing this!

**Comment by dovakin0007** [comment]
> Thanks for your patience on this! I think this PR looks ready to go, I pushed some code style nits to a different PR here: https://github.com/zed-industries/zed/pull/21082, so I'm going to close this one for now. But it's all the same code and commits, thank you for implementing this!

Thank you! ❤️

**Pull request closed without merging**